FLINK-30455 [core] Excluding java.lang.String and primitive types from closure cleaning

gunnarmorling opened the pull request:
https://issues.apache.org/jira/browse/FLINK-30455

## What is the purpose of the change

Avoiding reflective access to java.lang.String which fails on Java 17 and beyond.

## Brief change log

Excluding java.lang.String and primitive types from closure cleaning

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? No
  - If yes, how is the feature documented? Not applicable

gunnarmorling commented:
// CC @rmetzger

rmetzger commented:
Thanks a lot for opening a PR. It would be nice to have a Unit test for this change in the `ClosureCleanerTest` which fails on Java 17 (because of the reflective access), but succeeds after the fix. Otherwise, the functionality you've added might be removed accidentally (even though, admittedly, we'll have only test coverage on J17

rmetzger commented:
The commit message doesn't adhere to Flink's commit message pattern.
It needs to be
`[FLINK-30455] Excluding java.lang.String and primitive types ...` (e.g. Jira Id in brackets)

gunnarmorling commented:
Thanks a lot for reviewing @rmetzger; I have

* Reworked the commit message
* Added a test which would fail on JDK 17 without the fix; with the fix, it passes on 11 and 17

rmetzger commented:
Flink's brutally strict checkstyle is unhappy:
```
[INFO] There is 1 error reported by Checkstyle 8.14 with /tools/maven/checkstyle.xml ruleset.
[ERROR] src/test/java/org/apache/flink/api/java/ClosureCleanerTest.java:[243] (javadoc) JavadocStyle: First sentence should end with a period.
```

gunnarmorling commented:
Doh, good catch, sorry for not checking this locally. Added the missing "." and force-pushed.

gunnarmorling commented:
@rmetzger, how/where did you see this failure btw.? It seems the CI system isn't pushing any status updates back to this repo?

rmetzger commented:
Thanks for the fix.
You have to follow these links:
<img width="686" alt="image" src="https://user-images.githubusercontent.com/89049/209200325-8a1e737d-a722-4cc2-b42a-6585b8456e62.png">

zentol commented:
I'm curious what the endgame of this approach is supposed to be. Sure, fixing it for Strings isn't _bad_, but it's still a problem for virtually every other JDK type; BigDecimals, list/map/set implementations etc etc., or of any other modules that users would like to use in the future.

gunnarmorling commented:
Yeah, fair question; I'm not sure how useful that recursive cleaning approach is to begin with, it seems rather unlikely to encounter anonymous classes deeper in an object tree. `ClosureCleanerLevel.TOP_LEVEL` might be the more practical default behavior in that light. Other than that, traversing into _any_ JDK class probably shouldn't be attempted; arguably, it's a programming failure of the author of any serializable class if they allow for anonymous values in their fields.

I'd say though let's not get the perfect into the way of the good. As you say, this change unlocks a number of usages, so it's an improvement over the status quo. Happy to log a follow-up issue to discuss further along the lines above. WDYT?

zentol commented:
> As you say, this change unlocks a number of usages, so it's an improvement over the status quo.

I'm not even sure if it even _improves_ something. AFAIK this is just a warning but it doesn't impact anything. At most it should replace one error message with another one; some other illegal attempt reflection, of which there are plenty, including 3rd party libraries.

AFAICT this is something we'll just have to live with for the foreseeable future, so I'd rather not invest any resources into it for the time being.

gunnarmorling commented:
> AFAIK this is just a warning but it doesn't impact anything.

It causes the job to fail on Java 17 (where illegal reflective access is an exception by default now). So yes, this change helps with getting certain jobs running which otherwise would fail.

zentol commented:
That is because FLINK-25002 wasn't set up yet.

gunnarmorling commented:
I'd consider using `--add-opens` only a measure of last resort, it shouldn't be required whenever possible. Even if you set it in the Flink launch scripts, e.g. a user running Flink in their IDE would have to a) be aware of that requirement and b) know how to set it.

I feel you are pushing back on this change, and I would like to understand the reasons for this. In my opinion, while it's not the ultimate solution to the problem (maybe there is even none), it arguably does improve the situation for some scenarios, and I don't see a downside to it really. If you have any concerns, could you help me understand them? Thanks!
